### ISSUE TITLE
Categorical variables need to be hashable

### ISSUE THREAD

**Issue by erpheus** [scikit-optimize/scikit-optimize#681]
I've have a scikit-learn compatible model that receives some pretty complex parameters.

For example, if implementing a neural network it can receive the layers definition as a list of sizes, the list has the number of neurons in each layer (I know this would be hard to optimize with skopt and maybe a different approach would be better, but imagine it has to be like this for the sake of the argument). When I try to give BayesSearchCV a categorical search space dimension with each category corresponding to a different neural network architecture I get the following error:

```
TypeError: unhashable type: 'list' 
 
Traceback (most recent call last):
...REDACTED...
  File \"/opt/conda/lib/python3.6/site-packages/skopt/space/space.py\", line 423, in __init__
    self.transformer.fit(self.categories)
  File \"/opt/conda/lib/python3.6/site-packages/skopt/space/transformers.py\", line 53, in fit
    self.mapping_ = {v: i for i, v in enumerate(X)}
  File \"/opt/conda/lib/python3.6/site-packages/skopt/space/transformers.py\", line 53, in <dictcomp>
    self.mapping_ = {v: i for i, v in enumerate(X)}
TypeError: unhashable type: 'list'
```

I've looked at the code and I understand the need for the CategoricalEncoder to then use a one-hot encoding and the quite-standard implementation in skopt/space/transformers.py works great for many cases but not for unhashable types.

I propose relaxing the constraint that categories in a categorical search space need to be hashable to just having an equality relation defined. I realize that such an implementation would make the transform function slower but it is barely called in skopt so I think it could be a good tradeoff. Moreover, that sort of implementation could be a fallback in case hashing doesn't work.

What do you think?

I might have time soon to make a sample pull request.

**Comment by betatim** [comment]
The hashable constraint comes from Python's dictionary which we use to [map from index to actual value and back](https://github.com/scikit-optimize/scikit-optimize/blob/1c4c0f12ad8c1fe0a33542108fc0e55164138f9d/skopt/space/transformers.py#L53-L54). We'd have to rewrite this with a custom data structure that allows unhashable types.

A workaround could be using tuples instead of lists in your complex types. Tuples are immutable which means they are hashable (and can be used as key in a dictionary).

**Comment by erpheus** [comment]
Yes, rewritting with a custom data structure is what I’m proposing. And maybe only use it when types are not hashable while raising a warning.

I’m using tuples now as the library underneath supports it, but it is a bit of a pain as my data comes from a json file and I have to recursively convert nested lists to nested tuples. If it were just this usecase I’d agree to leave my workaround but I think there could be other usecases in which this problem would come up.

For example, a categorical dimension could be a list of custom loss functions, or a list of cross-validation objects, or any other thing not hashable and not easily convertible to a hashable type. Not to mention that in some cases the underlying estimators would have to be wrapped to undo the conversion (or some other mehtod such as gp_minimize would have to be used).

All in all, I think the restriction, although enforced by python’s dict class, is too strong. And I’m willing to try build an implementation if you think you would accept such a thing.

Thanks!

**Comment by betatim** [comment]
Hard to decide if this is a change to accept or not. My personal starting point is that a lot of thinking has gone into `dict` and its design. There are a lot of corner cases to handle/think about to create a replacement that works everywhere a `dict` works _and_ works for additional use cases. Combined with the experience that the complexity that comes with making something more general grows exponentially.

So without a proposed solution I would say writing individual small bits of glue for the cases where code really insists on a list instead of a tuple and you know that it is safe to convert back and forth is the way to go. Less exciting but overall quicker.

As an example of why you can't use a `list` as key:
```python
my_key = [1, 2, 3]
clever_dict = {my_key: 'abc'}
my_key[1] = 42
print(clever_dict.keys())
```
what _should_ the last line print? I am not sure I know a good answer to that question.

**Comment by janosh** [comment]
Having the same problem. Surely there are quite a number of use cases that require non-hashable `Categorical` values, making this a very useful feature.

**Comment by datason** [comment]
Just one another example when it's required non-hashable 'Categorical` values:

`### Search spaces 
vec_search = {
    'union__vec': Categorical([CountVectorizer(), TfidfVectorizer()]),
    'union__vec__tokenizer': Categorical([nltk_tokenize, None]), 
    'union__transformer_weights': Categorical(({'vec': 1.0, 'text_stats': 1.0},
                                                                         {'vec': 2.0, 'text_stats': 1.0}))
}`
Last line raises `TypeError: unhashable type: 'dict'`.
Here, I want to use BayesSearchCV for tuning `transformer_weights` in `sklearn.compose.ColumnTransformer`.

**Comment by alfonsomhc** [comment]
In case this is useful to someone: I had issues with the parameters for scikit learn's vectorizers. F.eks. HashingVectorizer takes a n_gram_range parameter of type tuple, and this isn't supported by skopt. My solution was to subclass the vectorizer so that it could take strings, which I then convert to tuples. It's important to understand that scikit learn estimators must not play with the input parameters in the init method, but you are free to do that in the fit method:

```
class HashingVectorizerCustom(HashingVectorizer):
    def fit(self, X, y=None):
        if isinstance(self.ngram_range, str):
            self.ngram_range = eval(self.ngram_range)
        return super().fit(X,y)
```

**Comment by tsminhuang** [comment]
Recently, I faced the same issue.
I want to search different class_wieght for a classifier.
I find one workaround is defined HashableDIct class yourself
There a lot of ways to do it.
https://stackoverflow.com/questions/1151658/python-hashable-dicts

```
class HashableDict(dict):
    def __hash__(self):
        return hash(tuple(sorted(self.items())))

## search space
 sgd_search_params = {
   'clf__loss': Categorical(['log', 'modified_huber']),
   'clf__class_weight': Categorical([HashableDict({0: 0.3, 1: 0}), HashableDict({0: 0.8, 1: 0}))
}
##
```

**Comment by pfuhlert** [comment]
I ran into the same issue and solved it by creating a `CategoricalList` to extend the `Categorical` class. This way I can use something like `[[1, 2, 3], [ 1, 4, 16], [1, 5, 25]]` to tune for one of the three sublists. 

Parameters can be defined like this:
```python
htune_params = {
    'mlp_architecture': CategoricalList([[32, 16], [16]]),
}
```

This is my implementation of `CategoricalList`:
```python
from skopt.space import Categorical

class CategoricalList(Categorical):
    def __init__(self, list_):
        super().__init__(self._convert_hashable(list_))        
        
    def _convert_hashable(self, list_of_lists):
        return [self._HashableListAsDict(list_) 
                for list_ in list_of_lists]

    class _HashableListAsDict(dict):
        def __init__(self, arr):
            self.update({i:val for i, val in enumerate(arr)})

        def __hash__(self):
            return hash(tuple(sorted(self.items())))   

        def __repr__(self):
            return str(list(self.values()))
```
